File: Lab5/TreeNode.py
class TreeNode(object):
     def __init__(self, val=0, left=None, right=None):
         self.val = val
         self.left = left
         self.right = right

def isValidBST(self, root):
     def helper(n, l=None, u=None):
          if n == None:
               return True
          if l != None and n.val <= l:
               return False
          if u != None and n.val >= u:
               return False
          if not helper(n.right,n.val,u):
               return False
          if not helper(n.left,l,n.val):
               return False
          return True
     return helper(root)

# testing
n1 = TreeNode(1)
n2 = TreeNode(2)
n9 = TreeNode(9)
n3 = TreeNode(3, n1, None)
n8 = TreeNode(8, n2, n9)
n4 = TreeNode(4, n3, n8)

print(isValidBST(n4,n4))  # False

n5 = TreeNode(5)
n7 = TreeNode(7)
n6 = TreeNode(6, n5, n7)
print(isValidBST(n6,n6))  # True

File: Lab5/TreeNode.java
public class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    
    TreeNode() {
    }
    
    TreeNode(int val) {
    	this.val = val;
    }

    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
    
    // check and set upper and lower bound at each particular node
    public boolean isValidBST(TreeNode root) {
    	int min = root.val;
    	int max = root.val;
    	
    	if(root.val > max || root.val < min)
    		return false;
    	
    	if(root.left.val < max) {
    		max = root.left.val;
    		return isValidBST(root.left);
    	}
    	else if(root.right.val>min) {
    		min = root.right.val;
    	}
    		return false;
    }
    
    public static void main(String[] args) {
    	TreeNode n1 = new TreeNode(1);
    	TreeNode n2 = new TreeNode(2);
    	TreeNode n9 = new TreeNode(9);
    	TreeNode n3 = new TreeNode(3,n1,null);
    	TreeNode n8 = new TreeNode(3,n2,n9);
    	TreeNode n4 = new TreeNode(3,n3,n8);
    	
    	System.out.println(n4.isValidBST(n4)+"");
    	
    }
}

